File: frontend/src/lib/calendar.ts
import { apiClient } from "./api";

const BASE = "/api/v1/calendar";

export type EventType = "meeting" | "task" | "booking" | "production" | "maintenance" | "other";
export type EventStatus = "scheduled" | "completed" | "cancelled";
export type ParticipantRole = "organizer" | "attendee" | "optional";
export type ResponseStatus = "pending" | "accepted" | "declined";
export type ResourceType = "employee" | "room" | "vehicle" | "machine" | "external";
export type ResourceStatus = "active" | "inactive" | "maintenance";
export type BookingStatus = "reserved" | "confirmed" | "cancelled";
export type RecurrenceFrequency = "daily" | "weekly" | "monthly" | "custom";
export type CAIAgentType = "schedule_optimizer" | "conflict_resolver";
export type CAIRecStatus = "pending" | "acknowledged" | "actioned" | "dismissed";

export interface CalendarResource {
  resource_id: string;
  resource_type: ResourceType;
  name: string;
  capacity: number;
  location?: string;
  description?: string;
  availability_rules: Record<string, unknown>;
  status: ResourceStatus;
  created_at: string;
}

export interface Participant {
  participant_id: string;
  event_id: string;
  user_id?: string;
  participant_name?: string;
  participant_email?: string;
  role: ParticipantRole;
  response_status: ResponseStatus;
}

export interface ResourceBooking {
  booking_id: string;
  event_id?: string;
  resource_id: string;
  start_datetime: string;
  end_datetime: string;
  status: BookingStatus;
  notes?: string;
  booked_by?: string;
  created_at: string;
}

export interface CalendarEvent {
  event_id: string;
  event_title: string;
  event_type: EventType;
  start_datetime: string;
  end_datetime: string;
  all_day_flag: boolean;
  created_by?: string;
  description?: string;
  location?: string;
  status: EventStatus;
  source_module?: string;
  recurrence_id?: string;
  created_at: string;
  participants: Participant[];
  bookings: { booking_id: string; resource_id: string; status: BookingStatus; start_datetime?: string; end_datetime?: string }[];
}

export interface AvailabilitySlot {
  start: string;
  end: string;
}

export interface CAIRec {
  rec_id: string;
  agent_type: CAIAgentType;
  title: string;
  body: string;
  score?: number;
  status: CAIRecStatus;
  rec_metadata: Record<string, unknown>;
  created_at: string;
}

export const calendarApi = {
  listResources: (params?: { resource_type?: string; status?: string }) =>
    apiClient.get<CalendarResource[]>(`${BASE}/resources`, { params }).then(r => r.data),
  createResource: (data: object) =>
    apiClient.post<CalendarResource>(`${BASE}/resources`, data).then(r => r.data),
  updateResource: (id: string, data: object) =>
    apiClient.patch<CalendarResource>(`${BASE}/resources/${id}`, data).then(r => r.data),

  listEvents: (params?: { start?: string; end?: string; event_type?: string; status?: string }) =>
    apiClient.get<CalendarEvent[]>(`${BASE}/events`, { params }).then(r => r.data),
  createEvent: (data: object) =>
    apiClient.post<CalendarEvent>(`${BASE}/events`, data).then(r => r.data),
  getEvent: (id: string) =>
    apiClient.get<CalendarEvent>(`${BASE}/events/${id}`).then(r => r.data),
  updateEvent: (id: string, data: object) =>
    apiClient.patch<CalendarEvent>(`${BASE}/events/${id}`, data).then(r => r.data),
  cancelEvent: (id: string) =>
    apiClient.delete<{ ok: boolean }>(`${BASE}/events/${id}`).then(r => r.data),

  listBookings: (params?: { resource_id?: string; start?: string; end?: string; status?: string }) =>
    apiClient.get<ResourceBooking[]>(`${BASE}/bookings`, { params }).then(r => r.data),
  createBooking: (data: object) =>
    apiClient.post<ResourceBooking>(`${BASE}/bookings`, data).then(r => r.data),
  confirmBooking: (id: string) =>
    apiClient.post<ResourceBooking>(`${BASE}/bookings/${id}/confirm`).then(r => r.data),
  cancelBooking: (id: string) =>
    apiClient.delete<{ ok: boolean }>(`${BASE}/bookings/${id}`).then(r => r.data),

  checkAvailability: (data: object) =>
    apiClient.post<{ available: boolean; conflicts: number }>(`${BASE}/availability/check`, data).then(r => r.data),
  getAvailableSlots: (data: object) =>
    apiClient.post<AvailabilitySlot[]>(`${BASE}/availability/slots`, data).then(r => r.data),

  listAIRecs: () =>
    apiClient.get<CAIRec[]>(`${BASE}/ai/recs`).then(r => r.data),
  runScheduleOptimizer: () =>
    apiClient.post<{ generated: number }>(`${BASE}/ai/run/schedule-optimizer`).then(r => r.data),
  runConflictResolver: () =>
    apiClient.post<{ generated: number }>(`${BASE}/ai/run/conflict-resolver`).then(r => r.data),
  ackAIRec: (id: string, data: { status: CAIRecStatus }) =>
    apiClient.patch<CAIRec>(`${BASE}/ai/recs/${id}`, data).then(r => r.data),
};

export const EVENT_COLOR: Record<EventType, string> = {
  meeting: "bg-blue-500",
  task: "bg-purple-500",
  booking: "bg-green-500",
  production: "bg-orange-500",
  maintenance: "bg-red-500",
  other: "bg-gray-400",
};

export const EVENT_BADGE: Record<EventType, string> = {
  meeting: "bg-blue-100 text-blue-700",
  task: "bg-purple-100 text-purple-700",
  booking: "bg-green-100 text-green-700",
  production: "bg-orange-100 text-orange-700",
  maintenance: "bg-red-100 text-red-700",
  other: "bg-gray-100 text-gray-600",
};

export const RESOURCE_ICON: Record<ResourceType, string> = {
  employee: "👤",
  room: "🏠",
  vehicle: "🚗",
  machine: "⚙️",
  external: "🌐",
};

export const STATUS_BADGE: Record<EventStatus, string> = {
  scheduled: "bg-blue-100 text-blue-700",
  completed: "bg-green-100 text-green-700",
  cancelled: "bg-gray-100 text-gray-500",
};

File: backend/app/models/calendar.py
from __future__ import annotations
import enum
import uuid
from datetime import datetime
from sqlalchemy import (
    Column, String, DateTime, Boolean, Integer,
    ForeignKey, Text, Enum as SAEnum, JSON, Float,
)
from sqlalchemy.orm import relationship
from app.db.base import Base


class EventType(str, enum.Enum):
    MEETING = "meeting"
    TASK = "task"
    BOOKING = "booking"
    PRODUCTION = "production"
    MAINTENANCE = "maintenance"
    OTHER = "other"


class EventStatus(str, enum.Enum):
    SCHEDULED = "scheduled"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class ParticipantRole(str, enum.Enum):
    ORGANIZER = "organizer"
    ATTENDEE = "attendee"
    OPTIONAL = "optional"


class ResponseStatus(str, enum.Enum):
    PENDING = "pending"
    ACCEPTED = "accepted"
    DECLINED = "declined"


class ResourceType(str, enum.Enum):
    EMPLOYEE = "employee"
    ROOM = "room"
    VEHICLE = "vehicle"
    MACHINE = "machine"
    EXTERNAL = "external"


class ResourceStatus(str, enum.Enum):
    ACTIVE = "active"
    INACTIVE = "inactive"
    MAINTENANCE = "maintenance"


class BookingStatus(str, enum.Enum):
    RESERVED = "reserved"
    CONFIRMED = "confirmed"
    CANCELLED = "cancelled"


class RecurrenceFrequency(str, enum.Enum):
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    CUSTOM = "custom"


class CAIAgentType(str, enum.Enum):
    SCHEDULE_OPTIMIZER = "schedule_optimizer"
    CONFLICT_RESOLVER = "conflict_resolver"


class CAIRecStatus(str, enum.Enum):
    PENDING = "pending"
    ACKNOWLEDGED = "acknowledged"
    ACTIONED = "actioned"
    DISMISSED = "dismissed"


class CalendarResource(Base):
    __tablename__ = "calendar_resources"

    resource_id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    resource_type = Column(SAEnum(ResourceType), nullable=False)
    name = Column(String(200), nullable=False)
    capacity = Column(Integer, default=1)
    location = Column(String(200))
    description = Column(Text)
    availability_rules = Column(JSON, default=dict)
    status = Column(SAEnum(ResourceStatus), default=ResourceStatus.ACTIVE)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    bookings = relationship("ResourceBooking", back_populates="resource")


class RecurrenceRule(Base):
    __tablename__ = "cal_recurrence_rules"

    recurrence_id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    frequency = Column(SAEnum(RecurrenceFrequency), nullable=False)
    interval = Column(Integer, default=1)
    end_date = Column(DateTime, nullable=True)
    max_occurrences = Column(Integer, nullable=True)
    days_of_week = Column(JSON, default=list)
    exceptions = Column(JSON, default=list)
    created_at = Column(DateTime, default=datetime.utcnow)

    events = relationship(
        "CalendarEvent", back_populates="recurrence",
        foreign_keys="CalendarEvent.recurrence_id",
    )


class CalendarEvent(Base):
    __tablename__ = "calendar_events"

    event_id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    event_title = Column(String(300), nullable=False)
    event_type = Column(SAEnum(EventType), default=EventType.MEETING)
    start_datetime = Column(DateTime, nullable=False)
    end_datetime = Column(DateTime, nullable=False)
    all_day_flag = Column(Boolean, default=False)
    created_by = Column(String(100))
    description = Column(Text)
    location = Column(String(200))
    status = Column(SAEnum(EventStatus), default=EventStatus.SCHEDULED)
    source_module = Column(String(50))
    source_ref_id = Column(String(36))
    recurrence_id = Column(
        String(36), ForeignKey("cal_recurrence_rules.recurrence_id"), nullable=True
    )
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    participants = relationship(
        "EventParticipant", back_populates="event", cascade="all, delete-orphan"
    )
    bookings = relationship(
        "ResourceBooking", back_populates="event", cascade="all, delete-orphan"
    )
    recurrence = relationship(
        "RecurrenceRule", back_populates="events", foreign_keys=[recurrence_id]
    )


class EventParticipant(Base):
    __tablename__ = "cal_event_participants"

    participant_id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    event_id = Column(
        String(36), ForeignKey("calendar_events.event_id", ondelete="CASCADE"), nullable=False
    )
    user_id = Column(String(36), nullable=True)
    participant_name = Column(String(200))
    participant_email = Column(String(200))
    role = Column(SAEnum(ParticipantRole), default=ParticipantRole.ATTENDEE)
    response_status = Column(SAEnum(ResponseStatus), default=ResponseStatus.PENDING)

    event = relationship("CalendarEvent", back_populates="participants")


class ResourceBooking(Base):
    __tablename__ = "cal_resource_bookings"

    booking_id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    event_id = Column(
        String(36), ForeignKey("calendar_events.event_id", ondelete="SET NULL"), nullable=True
    )
    resource_id = Column(
        String(36), ForeignKey("calendar_resources.resource_id"), nullable=False
    )
    start_datetime = Column(DateTime, nullable=False)
    end_datetime = Column(DateTime, nullable=False)
    status = Column(SAEnum(BookingStatus), default=BookingStatus.RESERVED)
    notes = Column(Text)
    booked_by = Column(String(100))
    created_at = Column(DateTime, default=datetime.utcnow)

    event = relationship("CalendarEvent", back_populates="bookings")
    resource = relationship("CalendarResource", back_populates="bookings")


class CAIRecommendation(Base):
    __tablename__ = "cal_ai_recommendations"

    rec_id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    agent_type = Column(SAEnum(CAIAgentType), nullable=False)
    title = Column(String(300))
    body = Column(Text)
    score = Column(Float, nullable=True)
    status = Column(SAEnum(CAIRecStatus), default=CAIRecStatus.PENDING)
    rec_metadata = Column(JSON, default=dict)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)


class ShiftType(str, enum.Enum):
    MORNING = "morning"
    AFTERNOON = "afternoon"
    NIGHT = "night"
    CUSTOM = "custom"


class ShiftSchedule(Base):
    """Staff shift scheduling — one row per user per date."""
    __tablename__ = "cal_shift_schedules"

    shift_id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(String(100), nullable=False, index=True)
    user_name = Column(String(200), nullable=True)
    department = Column(String(100), nullable=True, index=True)
    shift_date = Column(String(10), nullable=False, index=True)     # ISO date string
    shift_type = Column(SAEnum(ShiftType), nullable=False, default=ShiftType.MORNING)
    shift_start = Column(String(5), nullable=False)                 # HH:MM
    shift_end = Column(String(5), nullable=False)                   # HH:MM
    location = Column(String(200), nullable=True)
    approved_flag = Column(Boolean, default=False)
    approved_by = Column(String(200), nullable=True)
    notes = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
